src/Storages/ExternalDataSourceConfiguration.h

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <Parsers/ASTIdentifier.h>
+#include <Interpreters/Context.h>
+#include <Poco/Util/AbstractConfiguration.h>
+
+
+namespace DB
+{
+
+struct ExternalDataSourceConfiguration
+{
+    String host;
+    UInt16 port = 0;
+    String username;
+    String password;
+    String database;
+    String table;
+    String schema;
+
+    std::vector<std::pair<String, UInt16>> addresses; /// Failover replicas.
+
+    String toString() const;
+
+    void set(const ExternalDataSourceConfiguration & conf);
+};
+
+using ExternalDataSourceConfigurationPtr = std::shared_ptr<ExternalDataSourceConfiguration>;
+
+
+struct StoragePostgreSQLConfiguration : ExternalDataSourceConfiguration
+{
+    String on_conflict;
+};
+
+
+struct StorageMySQLConfiguration : ExternalDataSourceConfiguration
+{
+    bool replace_query = false;
+    String on_duplicate_clause;
+};
+
+struct StorageMongoDBConfiguration : ExternalDataSourceConfiguration
+{
+    String collection;
+    String options;
+};
+
+
+using EngineArgs = std::vector<std::pair<String, DB::Field>>;
```

Comment:
`EngineArgs` is a misleading type name, because it contains not all engine arguments. I'd suggest to call it `StorageSpecificArgs`.